spmd/compiler/distribute.py

```diff
@@ -206,38 +201,44 @@ def dummy_add(grad: torch.Tensor, zero: torch.Tensor) -> torch.Tensor:
 
     traced_dispatch.graph.lint()
 
-    return traced_dispatch
+    return traced_dispatch, node_to_obj[call_functions[0]]
 
 
 def _convert_output(
     gm: fx.GraphModule,
     node: fx.Node,
     node_to_obj: Dict[fx.Node, object],
-) -> None:
+) -> fx.Node:
     new_args = []
     has_partial = False
     for argument in node.args[0]:  # type: ignore
         if not isinstance(argument, fx.Node):
             new_args.append(argument)
             continue
-        obj = node_to_obj[argument]
 
-        has_partial = _is_partial_dtensor(obj)
+        obj = node_to_obj[argument]
 
-        if not has_partial:
+        if not _is_partial_dtensor(obj):
+            new_args.append(argument)
             continue
```

Comment:
This is the main fix . There's 2 issues

1) has_partial was set to True and then back to False if a Sharded() output appeared after a Partial_() output (I think this bug was added in the cleanup).

2) new_args needs to be appended with the output even if it's not partial. Thsi was being skipped -- and this is what caused less outputs than expected.

Reply:
"new_args needs to be appended with the output even if it's not partial." - Can you explain this a bit more? Not sure I got this issue.

Reply:
In this loop we are iterating through all the outputs.
We need to check, for each output, if they are Partial or not. 
- If they are Partial, we add the all reduce, and then append the output from all_reduce to the output list
- If they are not Partial, we should just append them to the new output list. But we were not doing that, so here's the bug.